src/gui/newwizard/setupwizardcontroller.cpp

```diff
@@ -1,409 +1,156 @@
     });
 
-    connect(_wizardWindow, &SetupWizardWindow::paginationEntryClicked, this, [this, paginationEntries](PageIndex currentPage, PageIndex clickedPageIndex) {
-        Q_ASSERT(currentPage < paginationEntries.size());
-        qCDebug(lcSetupWizardController) << "pagination entry clicked: current page" << currentPage << "clicked page" << clickedPageIndex;
-        nextStep(currentPage, clickedPageIndex);
+    connect(_context->window(), &SetupWizardWindow::paginationEntryClicked, this, [this](SetupWizardState clickedState) {
+        qCDebug(lcSetupWizardController) << "pagination entry clicked: current state" << _currentState << "clicked state" << clickedState;
+        changeStateTo(clickedState);
     });
-    connect(_wizardWindow, &SetupWizardWindow::nextButtonClicked, this, [this, paginationEntries](PageIndex currentPage) {
-        Q_ASSERT(currentPage < paginationEntries.size());
-        qCDebug(lcSetupWizardController) << "next button clicked on current page" << currentPage;
-        nextStep(currentPage, std::nullopt);
+
+    connect(_context->window(), &SetupWizardWindow::nextButtonClicked, this, [this]() {
+        qCDebug(lcSetupWizardController) << "next button clicked, current state" << _currentState;
+        _currentState->evaluatePage();
     });
 
     // in case the back button is clicked, the current page's data is dismissed, and the previous page should be shown
-    connect(_wizardWindow, &SetupWizardWindow::backButtonClicked, this, [this](PageIndex currentPage) {
-        // back button should be disabled on the first page
-        Q_ASSERT(currentPage > 0);
-        qCDebug(lcSetupWizardController) << "back button clicked on current page" << currentPage;
-        nextStep(currentPage, currentPage - 1);
+    connect(_context->window(), &SetupWizardWindow::backButtonClicked, this, [this]() {
+        // with enum classes, we have to explicitly cast to a numeric value
+        auto currentStateIdx = static_cast<int>(_currentState->state());
+        Q_ASSERT(currentStateIdx > 0);
+
+        qCDebug(lcSetupWizardController) << "back button clicked, current state" << _currentState;
+
+        auto previousState = static_cast<SetupWizardState>(currentStateIdx - 1);
+        changeStateTo(previousState);
     });
 }
 
 SetupWizardWindow *SetupWizardController::window()
 {
-    return _wizardWindow;
+    return _context->window();
 }
 
-void SetupWizardController::nextStep(std::optional<PageIndex> currentPage, std::optional<PageIndex> desiredPage)
+void SetupWizardController::changeStateTo(SetupWizardState nextState)
 {
     // should take care of cleaning up the page once the function has finished
-    QScopedPointer<AbstractSetupWizardPage> page(_currentPage);
+    QScopedPointer<AbstractSetupWizardState> page(_currentState);
 
     // initial state
-    if (!currentPage.has_value()) {
-        desiredPage = 0;
+    if (_currentState == nullptr) {
+        Q_ASSERT(nextState == SetupWizardState::FirstState);
     }
```

Comment:
Please move the check to the assert

Reply:
How? `Q_ASSERT(_currentState == nullptr && nextState == SetupWizardState::FirstState)` is not equal to the current code. `Q_ASSERT(_currentState == nullptr ? nextState == SetupWizardState::FirstState : true)` is less readable.

Reply:
`Q_ASSERT(_currentState == nullptr && nextState == SetupWizardState::FirstState) `
asserts the only valid combination and should be sufficient.

Reply:
I think you got this wrong. This is not the only valid combination. Initially `_currentState` is not set, it is initialized in this step. `FirstStep` is currently an alias of `ServerUrlState`.

Reply:
`Q_ASSERT(_currentState != nullptr || nextState == SetupWizardState::FirstState)`?